contexts/AssistantContext.tsx

```diff
@@ -193,49 +207,168 @@ export const AssistantProvider: React.FC<AssistantProviderProps> = ({ children }
 
     const isInConversationalMode = state.isConversationalMode
 
+    // t0: User taps stop
+    const t0 = Date.now()
+    console.log('[Metrics] t0: User stopped speaking')
+
     setState(prev => ({ ...prev, isListening: false, isProcessing: true }))
 
     try {
       const context = getContext()
 
-      // Stop recording and process
-      const result = await realtimeAgent.stopListeningAndProcess(messages, context)
+      // Stop recording and get transcription first (optimistic UI)
+      const audioUri = await realtimeAgent.stopRecordingAndGetAudio()
 
-      // In conversational mode: DON'T add messages to chat
-      // In chat mode: ADD messages to chat
-      if (!isInConversationalMode) {
-        const userMessage: Message = {
-          id: `user-${Date.now()}`,
-          role: 'user',
-          content: result.transcribedText,
-          timestamp: new Date(),
+      // t1: Audio file ready to upload
+      const t1 = Date.now()
+      console.log('[Metrics] t1: Audio file ready (t1-t0):', t1 - t0, 'ms')
+
+      // Transcribe immediately and show to user
+      // Pass last message as context for better accuracy
+      const previousMessage = messages.length > 0 ? messages[messages.length - 1].content : ''
+      const transcribePromise = realtimeAgent.transcribeAudio(
+        audioUri,
+        context.language === 'es' ? 'es' : 'en',
+        previousMessage
+      )
+
+      // Show transcription as soon as available (don't wait for LLM)
+      const transcribedText = await transcribePromise
```

Comment:
```suggestion
      const transcribedText = await realtimeAgent.transcribeAudio(
        audioUri,
        context.language === 'es' ? 'es' : 'en',
        previousMessage
      )
```

Which would be the difference with doing this?